### ISSUE TITLE
[TW-677] Erratic 'Task next' results

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#712]
_Ruthan Freese on 2011-04-29T00:27:52Z says:_ 

 > Front Desk(at)techshopfd ~ 
> $ task next 
> 
> ID Project .... 
> 
> ... 38 tasks, truncated to 21 tasks 

This is well and good.  Maybe still a few more tasks than I'd like, but I suppose it's what I get for not prioritizing or adding due dates to much of anything. 

Next, I add a task, or remove a task, or do anything at all, really.  Then: 

> Front Desk(at)techshopfd ~ 
> $ task next ... 
> 
> ... 38 tasks 

Short of restarting Cygwin, not even changing priorities and due dates yields anything except the whole list of tasks.

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2011-04-29T00:27:52Z
Modified: 2014-02-09T02:06:33Z
```

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-04-29T00:52:23Z says:_ 

 The 'next' report generates a list, with the 2 highest tasks from each of your projects.  Thus it seems you have 19 distinct projects.  This is functioning as intended.

I should say as _originally_ intended, because the next report is known to be of limited value.  This is why there has been a year or so of experimentation with a new next report, which will be appearing in taskwarrior 2.0.  It will not simply generate a list of high priority tasks, but will apply a tunable polynomial to gauge "urgency", and then sort by that.

A lot of work has gone into the new 'next' report.  Perhaps you have something to add?

http://tasktools.org/gitweb/gitweb.cgi?p=rfc.git;a=blob_plain;f=rfc31-urgency.txt;hb=HEAD
http://taskwarrior.org/boards/6/topics/1074
1. 707
1. 602
http://taskwarrior.org/boards/6/topics/995
http://taskwarrior.org/boards/8/topics/938
http://taskwarrior.org/projects/taskwarrior/wiki/Helper_Commands
http://taskwarrior.org/boards/6/topics/28

**Comment by taskwarrior** [comment]
_Ruthan Freese on 2011-04-29T01:33:53Z says:_ 

 I only have five projects, not counting the couple of times I've misspelled a project and then reassigned the task to the project that actually exists. 

It just doesn't seem to make sense for me to get 21 tasks once and 38 the next time.  

I'm exceedingly new to both TW and Using A Computer Like Someone Who Actually Knows What She's Doing, but I'll take a look at the links above and attempt to contribute in some wise!

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-04-29T01:59:32Z says:_ 

 I'm sorry, I was focusing on the '38', not the '21'.

The definition for the next report is to just display one page:

<pre>
task show report.next.filter

Config variable    Value                                  
report.next.filter status:pending limit:page depends.none:
</pre>

That 'limit:page' stops the display after one screenful of tasks.  Is it possible that your terminal window changed size from one run to the next?

If that's not it, then this is a bug.  Is it consistently inconsistent for you?

**Comment by taskwarrior** [comment]
_Ruthan Freese on 2011-04-29T02:23:44Z says:_ 

 Aha!  

I get the number of tasks equal to the number of monitor lines.  But there are five lines after that (two blank, one with the task count, and two for the prompt) and two above (for column headings plus a blank line) and I was indiscriminately expanding my monitor window to see the whole thing.  

It'd be handy if the limit took that into account, but it seems unlikely that (for instance) all prompts take up the same number of lines.  

And now I can probably modify config accordingly.  Thanks!

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-04-29T03:16:44Z says:_ 

 We're getting somewhere.  Let's keep digging.

Limit is supposed to take blank lines and task count into consideration, but that seems broken right now.

First, I read that you have a two-line prompt.  That's not something that taskwarrior expected, so there's one problem with limit:page.  Now bear in mind, you could always change that to "limit:20" or some such low value as a workaround.  But there's still some kind of problem.

Could you run "task diag", and either paste the results here, or mail them to paul(at)beckingham.net if you're not comfortable with that?  That will tell me if you're running Cygwin,  which version you are using, and it it's 1.9.4, how big taskwarrior thinks the terminal window is.  Could you also figure out how big the window is, in terms of width and height, in characters?  This will tell us whether taskwarrior is mistaken about that.

**Comment by taskwarrior** [comment]
_Ruthan Freese on 2011-04-29T03:42:13Z says:_ 

 <pre>
task 1.9.4
Platform: Cygwin 

Compiler
   Version: 4.3.4 20090804 (release) 1 
      Caps: _stdc _stdc_hosted +200112 +c1 +i4 +l4 +vp4 

Libraries
  Readline: n/a
       Lua: n/a

Build Features
     Built: Mar 5 2011 00:03:17
      Caps: +pthreads +srandom +random -uuid 

Configuration 
      File: /home/Front Desk/.taskrc (found), 1396 bytes, mode 100644
      Data: /home/Front Desk/.task (found), dir, mode 40755
   Locking: Enabled
     Regex: Disabled 

External Utilities 
       scp: n/a
     rsync: n/a
      curl: n/a

Tests
  UUID gen: 1000 unique UUIDs generated. 
     $TERM: cygwin (120x23)
</pre>


The window appears to actually only 80x19.*  A second run of task diag yields 51x12 with an actual size of 41x10, and a third run reports 120x14, actual size 80x11.  (Possibly task isn't accounting for lines hidden under scroll bars, but that doesn't explain the column discrepancy, as the horizontal scroll bar disappears at width 120.)  

(*) I guess this would explain the weird line breaks that accompany long tasks and/or annotations ... 

I'm also having problems editing report.next.filter -- either user error or a different bug?

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-04-29T03:52:48Z says:_ 

 Okay, you're up to date.  That's good for you, bad for me.  Please try this:

<pre>
export TERM=xterm-color
task diag
</pre>

No need to paste the results - I just want to see if the TERM value has an effect.  By the way, the scroll bars should have no effect.  Should.

What's happening with the filter?

Oh, and thanks for your patience.

**Comment by taskwarrior** [comment]
_Ruthan Freese on 2011-04-29T04:03:26Z says:_ 

 Doesn't seem to have had an effect -- diag returns 120x32 but is 80x26.  

As regards the filter, 
>task config report.next.filter 
returns
>No entry named 'report.next.filter' found. 
and 
>task config report.next.filter '' 
returns 
>Are you sure you want to add 'report.next.filter' with a value of ''? (y/n)
although 
>task show report.next.filter 
returns as you described above.

**Comment by taskwarrior** [comment]
_Ruthan Freese on 2011-04-29T04:04:22Z says:_ 

 Please excuse ugly formatting.  You're welcome to the patience.

**Comment by taskwarrior** [comment]
_Ruthan Freese on 2011-04-29T04:09:47Z says:_ 

 Found task _config; it includes report.next.filter.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-04-29T04:31:35Z says:_ 

 What terminal are you using for Cygwin?  Is it rxvt, the default or some other one?  Which version of Windows?  I'm looking to replicate this.

The config command is for *writing* the configuration value.   When you run:

<pre>
task config report.next.filter
...
</pre>

You are requesting that taskwarrior delete the filter, which doesn't exist in your .taskrc file.  Now assuming you have basically a default installation of taskwarrior, then you are using the default definition of next, which can be seen with:

<pre>
task show report.next

Config variable          Value                                                         
* ----------------------  ----------------------------------------------
report.next.columns      id,project,priority,due,active,age,description
report.next.description  Lists the most urgent tasks
report.next.filter       status:pending limit:page depends.none:
report.next.labels       ID,Project,Pri,Due,Active,Age,Description
report.next.sort         due+,priority-,active-,project+,description+
</pre>

To change the filter, you would need to enter:

<pre>
task config report.next.filter "some new value in quotes"
</pre>

This then creates an override in your ~/.taskrc file.

**Comment by taskwarrior** [comment]
_Ruthan Freese on 2011-05-02T21:04:36Z says:_ 

 Apologies for falling off the face of the earth over the weekend.  

echo $TERM tells me (rather unhelpfully) that I am using cygwin, but the cygwin directory has several rxvt* files in it, so it seems not unreasonable to assume that is what I am using.  

Windows XP Pro "Version 5.1 (Build 2600.xpsp_sp3_gdr.101209-1647 : Service Pack 3)", in fact.


As regards <pre>task config report.next.filter "limit:10"</pre>
, it also gets 

<pre>No entry named 'report.next.filter' found.</pre>

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-05-02T22:44:15Z says:_ 

 Please try this:

<pre>
TERM=xterm task diag
</pre>

Does it still get the size wrong?

Also, I think the "limit:10" is confusing the config command.  Please try:

<pre>
task config report.next.filter -- limit:10
</pre>

**Comment by taskwarrior** [comment]
_Ruthan Freese on 2011-05-02T23:08:59Z says:_ 

 The size is still wrong by about the same margin, it looks like (returns 120x31 for 80x25). 

The task config command works a-ok (having appended "status:pending depends.none:".)

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-05-08T22:58:11Z says:_ 

 This is being addressed in 2.0.0 by a brand new "next" report, that gauges urgency by menas of a tunable algorithm.

As I was obsoleting the code for the old next report, I noticed something odd - the code isn't hooked up.  That means there is no override that goes and finds urgent tasks from each project to populate this report, and it falls back to the same behavior as the "list" report.  You are right, Ruthan, the next report is broken.  My apologies.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-05-14T04:07:00Z says:_ 

 Obsolete - replaced by the new 'next' report, which works differently.
